data/transactions/logic/eval.go

```diff
@@ -264,19 +259,121 @@ type EvalParams struct {
 	// MinTealVersion is nil, we will compute it ourselves
 	MinTealVersion *uint64
 
-	// Amount "overpaid" by the top-level transactions of the
-	// group.  Often 0.  When positive, it is spent by application
-	// actions.  Shared value across a group's txns, so that it
-	// can be updated. nil is interpretted as 0.
+	// Amount "overpaid" by the transactions of the group.  Often 0.  When
+	// positive, it can be spent by inner transactions.  Shared across a group's
+	// txns, so that it can be updated (including upward, by overpaying inner
+	// transactions). nil is treated as 0 (used before fee pooling is enabled).
 	FeeCredit *uint64
 
 	Specials *transactions.SpecialAddresses
 
-	// determines eval mode: runModeSignature or runModeApplication
-	runModeFlags runMode
+	// Total pool of app call budget in a group transaction (nil before budget pooling enabled)
+	PooledApplicationBudget *int
+
+	// Total allowable inner txns in a group transaction (nil before inner pooling enabled)
+	pooledAllowedInners *int
+
+	// created contains resources that may be used for "created" - they need not be in
+	// a foreign array. They remain empty until createdResourcesVersion.
+	created *resources
+
+	// Caching these here means the hashes can be shared across the TxnGroup
+	// (and inners, because the cache is shared with the inner EvalParams)
+	appAddrCache map[basics.AppIndex]basics.Address
+
+	// Cache the txid hashing, but do *not* share this into inner EvalParams, as
+	// the key is just the index in the txgroup.
+	txidCache map[int]transactions.Txid
+
+	// The calling context, if this is an inner app call
+	caller *EvalContext
+}
+
+func copyWithClearAD(txgroup []transactions.SignedTxnWithAD) []transactions.SignedTxnWithAD {
+	copy := make([]transactions.SignedTxnWithAD, len(txgroup))
+	for i := range txgroup {
+		copy[i].SignedTxn = txgroup[i].SignedTxn
+		// leave copy[i].ApplyData clear
+	}
+	return copy
+}
+
+// NewEvalParams creates an EvalParams to use while evaluating a top-level txgroup
+func NewEvalParams(txgroup []transactions.SignedTxnWithAD, proto *config.ConsensusParams, specials *transactions.SpecialAddresses) *EvalParams {
+	apps := 0
+	for _, tx := range txgroup {
+		if tx.Txn.Type == protocol.ApplicationCallTx {
+			apps++
+		}
+	}
+
+	minTealVersion := ComputeMinTealVersion(txgroup, false)
+
+	var pooledApplicationBudget *int
+	var pooledAllowedInners *int
+
+	credit, _ := transactions.FeeCredit(txgroup, proto.MinTxnFee)
```

Comment:
please add a unit test for `FeeCredit` as well as test the returned error code.

Reply:
Added a unit test.

By the time we're here,  TxnGroupBatchVerify from verify/txn.go has run, which is essentially the same code, so it should not be possible for this to fail.  And the method has no way to exit with an error.

A couple options:
1. I could actually call FeeCredit from TxnGroupBatchVerify, so that it is more clear that this call will not fail.
2. I could return nil to the calller if FeeCredit fails here.
3. I could do both.

I like number 1 somewhat, though it means running through the txns twice. (I can't imagine that's a real problem, it's a tiny loop.) I don't like number 2 much as it just kicks the can down the road, and all callers ought to check for nil.

I don't know if that feels like enough for you - In general, what do we think is the right thing to do when code is running over a verified and/or "well formed" transactions in such a way that it knows certain problems can not occur?

Reply:
My guess, in fact, is that did use FeeCredit in the verification code at some point, and maybe I mismerged master at some point and reverted to doing the credit "by hand".  Let me know if you think that would solve this problem. It should be clearer that the txgroup passes the same criteria twice (and I can include a comment as such).

Reply:
Before I do this, could @tsachiherman comment on whether he thinks it's acceptable?  I would use `FeeCredit` in TxnBatchVerify, and then I would point that out here, as the explanation for why I need not check the error return.

Reply:
The fact that you need to do it from two places is a good indication that we haven't built the execution pipeline optimally here. I don't think it should be a blocker, but we should prepare for refactor this in a better way.
From practical perspective :
1. add a comment above the said call explaining the situation
2. Make TxnGroupBatchVerify call `transactions.FeeCredit`.
3. Create a ticket for scheduling that work

Reply:
This turned out to have a similar problem as we were discussing with ComputeMinTealVersion. Some code would like to work on the SignedTxns, and some on SignedTxnWithADs.  So I'm going to trim down the FeeCredit function to be local (`feeCredit`) have it operate on the WithADS, and not return an error (it will document that it depends upon the group being validated earlier.